### PR TITLE
add gke sink

### PR DESCRIPTION
**Pull request by aleksandra-malinowska**
Add GKE sink to export metrics directly to Stackdriver.

### CODE REVIEW

File: metrics/sinks/gke/gke.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	"github.com/golang/glog"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	stackdriver "google.golang.org/api/monitoring/v3"
+	gce_util "k8s.io/heapster/common/gce"
+	"k8s.io/heapster/metrics/core"
+)
+
+const (
+	maxTimeseriesPerRequest = 20
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: Also this can be shared with gcm sink

File: metrics/sinks/gke/gke.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	project           string
+	zone              string
+	stackdriverClient *stackdriver.Service
+}
+
+type metricMetadata struct {
+	MetricKind, ValueType, Name string
+}
+
+var (
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: const

File: metrics/sinks/gke/gke.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+)
+
+type gkeSink struct {
+	project           string
+	zone              string
+	stackdriverClient *stackdriver.Service
+}
+
+type metricMetadata struct {
+	MetricKind, ValueType, Name string
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: Please have one field per line.

File: metrics/sinks/gke/gke.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+			}
+			if len(req.TimeSeries) >= maxTimeseriesPerRequest {
+				sink.sendRequest(req)
+				req = getReq()
+			}
+		}
+	}
+}
+
+func CreateGKESink(uri *url.URL) (core.DataSink, error) {
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: This can be shared with gcm sink

File: metrics/sinks/gke/gke.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+		zone:              zone,
+		stackdriverClient: stackdriverClient,
+	}
+
+	glog.Infof("Created GKE sink")
+
+	return sink, nil
+}
+
+func (sink *gkeSink) sendRequest(req *stackdriver.CreateTimeSeriesRequest) {
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: this can be shared with gcm sink

File: metrics/sinks/gke/gke.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+		glog.V(4).Infof("Successfully sent %v timeseries to Stackdriver, project %v", len(req.TimeSeries), sink.project)
+	}
+}
+
+func (sink *gkeSink) translateMetric(timestamp time.Time, labels map[string]string, name string, value core.MetricValue, createTime time.Time) *stackdriver.TimeSeries {
+	switch name {
+	case "uptime":
+		point := sink.uptimePoint(timestamp, createTime, value)
+		resourceLabels := sink.getResourceLabels(labels)
+		//glog.Infof("Uptime for container: %v", resourceLabels["container_name"])
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: remove debug code

File: metrics/sinks/gke/gke.go [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+	}
+}
+
+func (sink *gkeSink) getResourceLabels(labels map[string]string) map[string]string {
+	return map[string]string{
+		"project_id":     sink.project,
+		"cluster_name":   "",
+		"zone":           sink.zone,
+		"instance_id":    labels[core.LabelHostID.Key],
+		"namespace_id":   "",
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: Why namespace_id is empty?

File: metrics/sinks/gke/gke.go [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+			StartTime: createTime.Format(time.RFC3339),
+		},
+		Value: &stackdriver.TypedValue{
+			DoubleValue:     float64(value.IntValue) / float64(time.Second/time.Millisecond),
+			ForceSendFields: []string{"DoubleValue"},
+		},
+	}
+}
+
+func fullProjectName(name string) string {
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: this can be shared with gcm sink

File: metrics/sinks/gke/gke.go [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	return fmt.Sprintf("projects/%s", name)
+}
+
+func getReq() *stackdriver.CreateTimeSeriesRequest {
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: this can be shared with gcm sink

File: metrics/sinks/gke/gke.go [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+	if err != nil {
+		glog.Errorf("Error while sending request to Stackdriver %v", err)
+	} else {
+		glog.V(4).Infof("Successfully sent %v timeseries to Stackdriver, project %v", len(req.TimeSeries), sink.project)
+	}
+}
+
+func (sink *gkeSink) translateMetric(timestamp time.Time, labels map[string]string, name string, value core.MetricValue, createTime time.Time) *stackdriver.TimeSeries {
+	switch name {
+	case "uptime":
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: Please use [MetricUptime.MetricDescriptor.Name](https://github.com/kubernetes/heapster/blob/master/metrics/core/metrics.go#L157)

File: metrics/sinks/gke/gke.go [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+		ValueType:  metadata.ValueType,
+		Resource: &stackdriver.MonitoredResource{
+			Labels: resourceLabels,
+			Type:   "gke_container",
+		},
+		Points: []*stackdriver.Point{point},
+	}
+}
+
+func (sink *gkeSink) uptimePoint(timestamp time.Time, createTime time.Time, value core.MetricValue) *stackdriver.Point {
```
Comment (piosz) [inline metrics/sinks/gke/gke.go]: Will you have a separate method for every point? At least interval can be somehow shared.

**Review by piosz — CHANGES_REQUESTED** [review]

File: metrics/sinks/factory.go [hunk 12]
```diff
@@ -44,6 +45,8 @@ func (this *SinkFactory) Build(uri flags.Uri) (core.DataSink, error) {
 		return elasticsearch.NewElasticSearchSink(&uri.Val)
 	case "gcm":
 		return gcm.CreateGCMSink(&uri.Val)
+	case "gke":
```
Comment (aleksandra-malinowska) [inline metrics/sinks/factory.go]: Done

File: metrics/sinks/gke/gke.go (hunk 9 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: To be refactored in follow-up PR

File: metrics/sinks/gke/gke.go (hunk 8 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: To be refactored in follow-up PR

File: metrics/sinks/gke/gke.go (hunk 11 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: Will change in next PR (adding more metrics)

File: metrics/sinks/gke/gke.go (hunk 7 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: Fixed

File: metrics/sinks/gke/gke.go (hunk 6 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: Done

File: metrics/sinks/gke/gke.go (hunk 10 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: Done

File: metrics/sinks/gke/gke.go (hunk 3 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: Done

File: metrics/sinks/gke/gke.go (hunk 4 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: To be refactored in follow-up PR

File: metrics/sinks/gke/gke.go (hunk 5 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: To be refactored in follow-up PR

File: metrics/sinks/gke/gke.go (hunk 1 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: Done

File: metrics/sinks/gke/gke.go (hunk 2 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/gke/gke.go]: AFAIU it's impossible to declare const structs?

File: metrics/sinks/stackdriver/stackdriver.go [hunk 13]
```diff
@@ -0,0 +1,190 @@
+// Copyright 2015 Google Inc. All Rights Reserved.
```
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: 2017

File: metrics/sinks/stackdriver/stackdriver.go [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+
+func (sink *stackdriverSink) uptimePoint(timestamp time.Time, createTime time.Time, value core.MetricValue) *sd_api.Point {
+	return &sd_api.Point{
+		Interval: &sd_api.TimeInterval{
+			EndTime:   timestamp.Format(time.RFC3339),
+			StartTime: createTime.Format(time.RFC3339),
+		},
+		Value: &sd_api.TypedValue{
+			DoubleValue:     float64(value.IntValue) / float64(time.Second/time.Millisecond),
+			ForceSendFields: []string{"DoubleValue"},
```
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: Why do you need this?

File: metrics/sinks/stackdriver/stackdriver.go [hunk 15]
```diff
@@ -0,0 +1,190 @@
...
+		"namespace_id":   labels[core.LabelPodNamespaceUID.Key],
+		"pod_id":         labels[core.LabelPodId.Key],
+		"container_name": labels[core.LabelContainerName.Key],
+	}
+}
+
+func createTimeSeries(resourceLabels map[string]string, labels map[string]string, metadata *metricMetadata, point *sd_api.Point) *sd_api.TimeSeries {
+	return &sd_api.TimeSeries{
+		Metric: &sd_api.Metric{
+			Labels: map[string]string{},
```
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: Why do you need to specify empty set of labels?

File: metrics/sinks/stackdriver/stackdriver.go [hunk 16]
```diff
@@ -0,0 +1,190 @@
...
+		"cluster_name":   "",
+		"zone":           sink.zone,
+		"instance_id":    labels[core.LabelHostID.Key],
+		"namespace_id":   labels[core.LabelPodNamespaceUID.Key],
+		"pod_id":         labels[core.LabelPodId.Key],
+		"container_name": labels[core.LabelContainerName.Key],
+	}
+}
+
+func createTimeSeries(resourceLabels map[string]string, labels map[string]string, metadata *metricMetadata, point *sd_api.Point) *sd_api.TimeSeries {
```
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: labels are not used

File: metrics/sinks/stackdriver/stackdriver.go [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+var (
+	uptimeMD = &metricMetadata{
+		MetricKind: "CUMULATIVE",
+		ValueType:  "DOUBLE",
+		Name:       "container.googleapis.com/container/uptime",
+	}
+)
+
+func (sink *stackdriverSink) Name() string {
+	return "GKE Sink"
```
Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: Stackdriver sink

**Review by piosz — CHANGES_REQUESTED** [review]

File: metrics/sinks/stackdriver/stackdriver.go (hunk 14 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/stackdriver/stackdriver.go]: Not adding this field caused custom metrics to be rejected by Stackdriver. I did a test and writing GKE container metrics seems to work without it, removing this (although kubelet-to-gcm still uses it?)

File: metrics/sinks/stackdriver/stackdriver.go (hunk 15 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/stackdriver/stackdriver.go]: Removed

File: metrics/sinks/stackdriver/stackdriver.go (hunk 16 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/stackdriver/stackdriver.go]: Fixed

File: metrics/sinks/stackdriver/stackdriver.go (hunk 17 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/stackdriver/stackdriver.go]: Fixed

File: metrics/sinks/stackdriver/stackdriver.go (hunk 13 above)
Comment (aleksandra-malinowska) [inline metrics/sinks/stackdriver/stackdriver.go]: Fixed

Comment (aleksandra-malinowska) [inline metrics/sinks/stackdriver/stackdriver.go]: This causes 'make sanitize' to fail, so keeping 2015

Comment (piosz) [inline metrics/sinks/stackdriver/stackdriver.go]: ACK

**Review by piosz — APPROVED** [review]
LGTM, thanks for the PR!

**Comment by DirectXMan12** [comment]
You'll need to also update the sink owners file.

**Pull request merged**